source/elastic_solver.py

```diff
@@ -204,52 +336,80 @@ def _setup_boundary_conditions(self):
+    
+    def _setup_function_spaces(self):
+        """
+        Class method setting up function spaces.
+        """
+        assert hasattr(self, "_mesh")
+        cell = self._mesh.ufl_cell()
+        # element formulation
+        elemU = dlfn.VectorElement("CG", cell, self._p_deg)
+        # mixed function space
+        self._Vh = dlfn.FunctionSpace(self._mesh, elemU)
+        self._n_dofs = self._Vh.dim()
+        # print info
+        assert hasattr(self, "_n_cells")
+        dlfn.info("Number of cells {0}, number of DoFs: {1}".format(self._n_cells, self._n_dofs))
+        
+    def set_dimensionless_numbers(self, C, D=None):
+        """
+        Updates the parameters of the model by creating or modifying class
+        objects.
+        """
+        assert isinstance(C, float)
+        assert isfinite(C)
+        assert C > 0.0
+        if not hasattr(self, "_C"):
+            self._C = dlfn.Constant(C)
+        else:
+            self._C.assign(C)
 
+        if D is not None:
+            assert isinstance(D, float)
+            assert isfinite(D)
+            assert D > 0.0
+            if not hasattr(self, "_D"):
+                self._D = dlfn.Constant(D)
+            else:
+                self._D.assign(D)
+    
     def _setup_problem(self):
         """
-        Method setting up non-linear solver objects of the stationary problem.
+        Method setting up nonlinear solver objects of the stationary problem.
         """
         assert hasattr(self, "_mesh")
         assert hasattr(self, "_boundary_markers")
 
         self._setup_function_spaces()
         self._setup_boundary_conditions()
 
-        # creating test and trial functions
-        u = dlfn.TrialFunction(self._Vh)
-        v = dlfn.TestFunction(self._Vh)
+        # creating test function
+        self._v = dlfn.TestFunction(self._Vh)
+        self._u = dlfn.TrialFunction(self._Vh)
 
         # solution
         self._solution = dlfn.Function(self._Vh)
 
         # volume element
-        dV = dlfn.Measure("dx", domain=self._mesh)
-        dA = dlfn.Measure("ds", domain=self._mesh, subdomain_data=self._boundary_markers)
-
-        # dimensionless parameters
-        C = self._C
+        self._dV = dlfn.Measure("dx", domain=self._mesh)
+        self._dA = dlfn.Measure("ds", domain=self._mesh, subdomain_data=self._boundary_markers)
         
-        # auxiliary function
-        def sym_grad(u):
-            return dlfn.Constant(0.5) * (grad(u) + grad(u).T)
+        # setup the parameters for the elastic law
+        self._elastic_law.set_parameters(self._mesh, self._C, self._u, self._v, self._solution)
         
-        # weak forms
-        # virtual work of internal forces
-        strain = sym_grad(u)
-        dstrain = sym_grad(v)
-        dw_int = (C * dlfn.tr(strain) * dlfn.tr(dstrain)
-                  + inner(dlfn.Constant(2.0) * strain, dstrain)
-                  ) * dV
+        # virtual work
+        self._dw_int = self._elastic_law.dw_int()
```

Comment:
`self._elastic_law.dw_int()` could also simply return the virtual energy density as function of the `solution` and the `test_function`. Then you could write
```python3
# setup the parameters for the elastic law
self._elastic_law.set_parameters(elasticity_ratio)
# virtual work
self._dw_int = self._elastic_law.dw_int(solution, test_function) * dV
```